File: java-db-sample/src/main/java/com/example/java/dto/UserDto.java
package com.example.java.dto;

import java.util.Objects;

/**
 * Java版のDTO（Data Transfer Object）
 *
 * 特徴:
 * - private フィールドとgetter/setterを持つPOJO
 * - equals(), hashCode(), toString()を手動で実装
 * - ボイラープレートコードが多い
 * - イミュータブルにするには全フィールドをfinalにし、setterを削除する必要がある
 * - nullableなフィールド（nickname, phoneNumber, address）に対して、
 *   使用時に常にnullチェックが必要（NullPointerExceptionのリスク）
 */
public class UserDto {
    private Long id;
    private String name;
    private String email;
    private int age;
    // Nullable fields (DBでNULLが許可されているカラム)
    private String nickname;      // ニックネーム（任意）
    private String phoneNumber;   // 電話番号（任意）
    private String address;       // 住所（任意）

    public UserDto() {
    }

    public UserDto(Long id, String name, String email, int age) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.age = age;
    }

    public UserDto(Long id, String name, String email, int age, String nickname, String phoneNumber, String address) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.age = age;
        this.nickname = nickname;
        this.phoneNumber = phoneNumber;
        this.address = address;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public String getNickname() {
        return nickname;
    }

    public void setNickname(String nickname) {
        this.nickname = nickname;
    }

    public String getPhoneNumber() {
        return phoneNumber;
    }

    public void setPhoneNumber(String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        UserDto userDto = (UserDto) o;
        return age == userDto.age &&
                Objects.equals(id, userDto.id) &&
                Objects.equals(name, userDto.name) &&
                Objects.equals(email, userDto.email) &&
                Objects.equals(nickname, userDto.nickname) &&
                Objects.equals(phoneNumber, userDto.phoneNumber) &&
                Objects.equals(address, userDto.address);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name, email, age, nickname, phoneNumber, address);
    }

    @Override
    public String toString() {
        return "UserDto{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", email='" + email + '\'' +
                ", age=" + age +
                ", nickname='" + nickname + '\'' +
                ", phoneNumber='" + phoneNumber + '\'' +
                ", address='" + address + '\'' +
                '}';
    }
}

File: kotlin-db-sample/src/main/kotlin/com/example/kotlin/dto/UserDto.kt
package com.example.kotlin.dto

/**
 * Kotlin版のDTO（Data Transfer Object）
 *
 * 特徴:
 * - data classを使用することで、equals(), hashCode(), toString(), copy()が自動生成される
 * - getter/setterは不要（プロパティアクセス）
 * - 1行で定義できる（ボイラープレートコードが激減）
 * - valを使うことで簡単にイミュータブルにできる
 * - デフォルト値を設定できる
 * - 名前付き引数でインスタンス生成が明確
 *
 * Null安全性:
 * - nickname, phoneNumber, addressは String? 型（nullableな型）
 * - 型システムレベルでnull安全が保証される
 * - Safe call演算子(?.)、Elvis演算子(?:)で安全にnullを扱える
 */
data class UserDto(
    val id: Long? = null,
    val name: String,
    val email: String,
    val age: Int,
    val nickname: String? = null,      // ニックネーム（任意）
    val phoneNumber: String? = null,   // 電話番号（任意）
    val address: String? = null        // 住所（任意）
)

/**
 * Java版と比較:
 * - Java版: 約130行（フィールド、コンストラクタ、getter/setter、equals、hashCode、toString）
 * - Kotlin版: 約15行（data classで全て自動生成）
 *
 * 使用例:
 * val user = UserDto(
 *     id = 1,
 *     name = "太郎",
 *     email = "taro@example.com",
 *     age = 30,
 *     nickname = "タロちゃん"  // nullableなフィールド
 * )
 * val copied = user.copy(age = 31)  // copyメソッドも自動生成
 *
 * Null安全な処理例:
 * val displayName = user.nickname ?: user.name  // nicknameがnullならnameを使用
 * val phoneLength = user.phoneNumber?.length ?: 0  // safe callでNullPointerExceptionを回避
 */
